modules/datastore/src/Storage/QueryFactory.php

```diff
@@ -75,6 +76,68 @@ public function populateQuery(): Query {
     return $query;
   }
 
+  /**
+   * Helper function for adding datastore groups to the given query.
+   *
+   * @param Drupal\common\Storage\Query $query
+   *   DKAN query object we're building.
+   *
+   * @throws \Exception
+   *   When ungrouped properties are found in the datastore query.
+   */
+  private function populateQueryGroupBy(Query $query): void {
+    $groups = $this->extractPropertyNames($this->datastoreQuery->{"$.groups"} ?? []);
+    if (empty($groups)) {
+      return;
+    }
+    $props = $this->extractPropertyNames($this->datastoreQuery->{"$.properties"} ?? []);
+    if ($ungrouped = array_diff($props, $groups)) {
+      throw new \Exception('Un-grouped properties found in aggregate query: ' . implode(', ', $ungrouped));
+    }
+    $query->groupby = $groups;
+  }
+
+  /**
+   * Extract list of property names from an array of properties.
+   *
+   * Properties can be either objects, associative-arrays, or strings. If a
+   * property is an object or associative-array, the property name is stored in
+   * the "property" field. If a property is a string, the string itself is a
+   * property name.
+   *
+   * @param array $props
+   *   List of query properties (which can be a string, associative array, or
+   *   object, see above).
+   *
+   * @return string[]
+   *   List of extracted property names.
+   *
+   * @throws \Exception
+   *   When an invaid property is encountered.
+   */
+  protected function extractPropertyNames(array $props): array {
+    return array_filter(array_map(function ($prop) {
+      if (is_string($prop)) {
+        return $prop;
+      }
+      // If prop is an array, convert it to an object.
+      if (is_array($prop)) {
+        $prop = (object) $prop;
+      }
+      if (is_object($prop)) {
+        if (isset($prop->property)) {
+          return $prop->property;
+        }
+        // If this property object is an expression, ignore it.
+        if (isset($prop->expression)) {
+          return NULL;
```

Comment:
In this case, wouldn't we want to return the expression's alias, not `null`?

Reply:
The name of this method is a bit confusing. I'm trying to extract "property properties" not "expression properties" :sweat_smile: